src/devices/src/virtio/queue.rs

```diff
@@ -393,6 +401,82 @@ impl Queue {
+        // We need to find the `used_event` field from the avail ring.
+        let used_event_addr = self
+            .avail_ring
+            .unchecked_add(u64::from(4 + 2 * self.actual_size()));
+
+        Wrapping(mem.read_obj::<u16>(used_event_addr).unwrap())
+    }
+
+    /// Helper method that writes `val` to the `avail_event` field of the used ring.
+    fn set_avail_event(&mut self, val: u16, mem: &GuestMemoryMmap) {
+        let avail_event_addr = self
+            .used_ring
+            .unchecked_add(u64::from(4 + 8 * self.actual_size()));
+
+        mem.write_obj(val, avail_event_addr).unwrap();
+    }
+
+    /// Try to enable notification events from the guest driver. Returns true if notifications were
+    /// successfully enabled. Otherwise it means that one or more descriptors can still be consumed
+    /// from the available ring and we can't guarantee that there will be a notification. In this
+    /// case the caller might want to consume the mentioned descriptors and call this method again.
+    pub fn try_enable_notification(&mut self, mem: &GuestMemoryMmap) -> bool {
+        // If the device doesn't use notification suppression, we'll continue to get notifications
+        // no matter what.
+        if !self.uses_notif_suppression {
+            return true;
+        }
+
+        if self.len(mem) != 0 {
+            return false;
+        }
+
+        // Set the next expected avail_idx as avail_event.
+        self.set_avail_event(self.next_avail.0, mem);
+
+        // Make sure all subsequent reads are performed after `set_avail_event`.
+        fence(Ordering::SeqCst);
+
+        // If the actual avail_idx is different than next_avail one or more descriptors can still
+        // be consumed from the available ring.
+        self.next_avail.0 == self.avail_idx(mem).0
+    }
+
+    /// Enable/disable notification suppression.
+    pub fn enable_notif_suppression(&mut self) {
+        self.uses_notif_suppression = true;
+    }
+
+    /// Check if we need to kick the guest.
+    ///
+    /// Please note this method has side effects: once it returns `true`, it considers the
+    /// driver will actually be notified, and won't return `true` again until the driver
+    /// updates `used_event` and/or the notification conditions hold once more.
+    ///
+    /// This is similar to the `vring_need_event()` method implemented by the Linux kernel.
+    pub fn prepare_kick(&mut self, mem: &GuestMemoryMmap) -> bool {
+        // If the device doesn't use notification suppression, always return true
+        if !self.uses_notif_suppression {
+            return true;
+        }
+
+        // We need to expose used array entries before checking the used_event.
+        fence(Ordering::SeqCst);
+
+        let new = self.next_used;
+        let old = self.next_used - self.num_added;
+        let used_event = self.used_event(mem);
+
+        self.num_added = Wrapping(0);
```

Comment:
just an idea: since this is the only state change (side effect) of this function, would it make sense to change this fn to `needs_kick()` which only does checking (with no side effects) and do `self.num_added = 0` in actual `do_kick()` logic?

Reply:
This is better from a naming point of view, but it introduces the possibility of forgetting to call `do_kick()` after `needds_kick()`. Actually the code was similar to this in an earlier version of the PR and when I tried to use this logic to implement notif suppression on net also in order to do some experiments I forgot to call `do_kick()` and it lead to freezes. I think it's safer to keep it as it is now.

Reply:
I still think current code is not ideal since it does only half of something. My preference was that we either:
- have single function which checks if IRQ needed then also kicks
- cleanly split checking and kicking

But I see that it's not easy to do that since part of the logic is in the queue while irq logic is in the device. I will not block the PR on this since it's a two-way door implementation detail.

Reply:
Yes, exactly. This is the first half of a kick as specified [here](https://elixir.bootlin.com/linux/v5.13/source/drivers/virtio/virtio_ring.c#L1814). It's also used in the same way in the [kernel block device](https://elixir.bootlin.com/linux/v5.13/source/drivers/block/virtio_blk.c#L210) btw. 

Another option is to have a method `Queue::kick(&self, irq_trigger: IrqTrigger)`. This way it could check if a notification is needed and call `irq_trigger.trigger_irq()` if so. But it would probably raise question related to rust-vmm upstream compatibility. Also the IrqTrigger is not mature enough and I'm not sure how it will evolve and if it will be so easy to pass it to the `Queue::kick()` in the future.

Persoanlly for the moment I think it would be safest to keep it as it is and maybe revisit this after migrating to the Queue implementation in rust-vmm.